Polyfill: support same-origin iframe targets

dvoytenko opened the pull request:
The polyfill is updated to support observing targets in the same-origin iframes. This is already a spec'd behavior and only polyfill abilities are enhanced. See https://www.w3.org/TR/intersection-observer/#dom-intersectionobserver-rootmargin for more info.

Notably, the changes to do not polyfill the `IntersectionObserver` inside an iframe, but simply allow observing using the top-level context `IntersectionObserver`. The final interface binding can be done by manually:

```
iframe.contentWindow.IntersectionObserver = window.IntersectionObserver;
iframe.contentWindow.IntersectionObserverEntry = window.IntersectionObserverEntry;
```

In practice, however, I find that even when same-origin iframes are used, the root observer is used for them. The tests show this pattern.

philipwalton commented on polyfill/intersection-observer.js:
```diff
@@ -693,6 +793,28 @@ function getEmptyRect() {
 
+
+/**
+ * Inverts the intersection and bonding rect from the parent (frame) BCR to
```
s/bonding/bounding/

dvoytenko replied:
done

philipwalton commented on polyfill/intersection-observer-test.html:
```diff
@@ -27,6 +27,14 @@
 
+  <!-- Force polyfill. -->
+  <script>
+    if (false) {
```
Did you mean to keep this here?

dvoytenko replied:
Yes, but I can remove if you don't want it here. As it is, I thought of this as a documentation for how to quickly test the polyfill on a modern browser.

philipwalton replied:
I think it's fine to keep it here (to explain how to test, as you say) but I'd remove the `if (false)` and replace it with a comment that says something along the lines of: "comment out the below script to test the default behavior" (or something like that).

dvoytenko replied:
Done.

dvoytenko commented:
> Thanks for filing this PR. These changes look pretty reasonable on first look (with a few nits). Have you looked into added a test for this new behavior?

Definitely. I added several tests in the `describe('same-origin iframe')`. I added tests, for three categories of things:
 1. Monitoring scrolling/mutations/etc for iframes. Similar to non-iframed case.
 2. Calculation of intersectionRect and other entry structs.
 3. Main-vs-iframe-context coordination tests.

philipwalton commented:
Ahh, sorry, the GitHub UI was hiding the diff for that file, so I missed it. Will take a look later today.

jridgewell commented on polyfill/intersection-observer.js:
```diff
@@ -262,58 +266,138 @@ IntersectionObserver.prototype._parseRootMargin = function(opt_rootMargin) {
-    removeEvent(document, 'scroll', this._checkForIntersections, true);
+  // Check if any dependent targets are still remaining.
+  var dependentTargets =
+      this._observationTargets.filter(function(item) {
```
We can use `_observationTargets.some()` to allow early exiting iteration.

dvoytenko replied:
Oops. Definitely!

jridgewell commented on polyfill/intersection-observer.js:
```diff
@@ -262,58 +266,138 @@ IntersectionObserver.prototype._parseRootMargin = function(opt_rootMargin) {
+ * @private
+ */
+IntersectionObserver.prototype._unmonitorAllIntersections = function() {
+  var unsubscribes = [].concat(this._monitoringUnsubscribes);
```
Nit: Use `_monitoringUnsubscribes.slice()`

philipwalton commented on polyfill/intersection-observer.js:
```diff
@@ -734,6 +861,21 @@ function getParentNode(node) {
 
 
+/**
+ * Returns the embedding framem element, if any.
```
s/framem/frame/

philipwalton commented on polyfill/intersection-observer-test.js:
```diff
@@ -941,6 +941,652 @@ describe('IntersectionObserver', function() {
 
   });
 
+  describe('same-origin iframe', function() {
```
Can you move these tests inside the `observe()` describe block. The structure of the tests currently uses a single describe block per method, so I think it makes sense to stick with that unless you think these tests really don't fit that pattern.

dvoytenko replied:
The tests are more orthogonal in this case and set up is very special for iframes and I wanted to reduce the spread across the board. But I also I wanted to make it concentrated to communicate "here are all the things that are special about iframes".

philipwalton replied:
OK. I still think it'd be nice if there were a way to logically group the same-document and iframe tests (maybe run all the tests in both scenarios), but since that would likely require a larger refactor, I don't want to block adding this feature.

philipwalton commented on polyfill/intersection-observer-test.js:
```diff
@@ -941,6 +941,652 @@ describe('IntersectionObserver', function() {
+        'body {margin: 0}' +
+        ' .target {height: 200px; margin-bottom: 2px; background: blue;}' +
+        '</style>';
+      return new Promise(function(resolve, reject) {
```
When this polyfill was written (years ago) it was designed to work all the way back to IE7.

Using `Promise` here would break the tests when run in IE, but I think you could get away with still using it by adding that polyfill to the polyfills the test runner loads: https://github.com/w3c/IntersectionObserver/blob/master/polyfill/intersection-observer-test.html#L22

dvoytenko replied:
Good point. I removed the promise. The rest of tests are using `done` callback, so this one should do as well.

philipwalton commented on polyfill/intersection-observer-test.js:
```diff
@@ -941,6 +941,652 @@ describe('IntersectionObserver', function() {
+      io.observe(iframeTargetEl2);
+    });
+
+    it('calculates rects for a fully visible frame', function(done) {
```
Can you confirm that this test is passing for you? I see failures when running it both with and without the polyfill. (Note: I see the same for the tests whose description starts with "calculates..."):

<img width="951" alt="Screen Shot 2020-04-08 at 5 44 12 PM" src="https://user-images.githubusercontent.com/326742/78846453-aeb4fd00-79c0-11ea-97ba-800e6c18de2c.png">

dvoytenko replied:
This is very odd. And especially w.r.t. to `right: 85`. But I can confirm that all tests pass for me for both native and polyfill and on all the browsers that I've tried. And I've tested on Chrome/Firefox/Safari. Having difficulties testing on IE at the moment, but will certainly find a way to test IE before merging (and any advise would be appreciated).

dvoytenko replied:
Turns out, this is not odd at all. Just the test wasn't accounting for the scrollbar. I fixed it. And tested on IE11 and IE10.

dvoytenko commented:
@philipwalton I fixed the tests and tested with IE11 and IE10. PTAL.

dvoytenko commented:
Close/reopen to run CI again.
